### DISCUSSION TITLE
Debug Language Server

### DISCUSSION THREAD

**Discussion by soerendomroes** [lf-lang/lingua-franca#1095] [Show and tell]
Hi,
I am trying to document a little bit how the language server can be debugged together with the VS Code.
First, you need to be able to connect your extension via a socket, similar to the implementation of the [KIELER VS Code extension](https://github.com/kieler/vscode/blob/main/keith-vscode/src/extension.ts) (see `createServerOptions` which is called by `activate`).
We set the corresponding environment variable in the launch config (of course KEITH_LS_PORT is only the corresponding variable name used in the `extension.ts`):
```
"env": {
    "KEITH_LS_PORT": "5007"
},
 ```
The server needs to be started on a socket. I only know how to do that in Eclipse but you just start your application. If I understand your `LanguageDiagramServer` you just call the `AbstractLanguageServer` in Klighd and let it handle the connection. This language server tries to read `-Dport`. By specifying `-Dport=5007` as a VM argument when starting your `LanguageDiagramServer` as a normal java application, the language server uses the same socket the VS extension connects to.

Hopefully, this is enough to get a debugging setup to work. I currently do not have time to do that myself but feel free to ask questions should any arise.

**Comment by soerendomroes** [comment]
If you want to debug the diagram web view you have to know the following:
This can only be done in the developer tools in the browser (since the webview is not and cannot be directly started from your developer VS Code).

klighd-core and klighd-interactive cannot be debugged in VS Code since they are not called by VS Code but by the launched VS Code at some point.
They need to be debugged in the developer tools of the launched VS Code.
You can find the code after opening the diagram in the `<random number directory>/active|pending frame/webpack-internal`. Based on your webpack configuration it is more or less readable.

**Comment by soerendomroes** [comment]
See https://github.com/lf-lang/vscode-lingua-franca/tree/sdo/debugging for connecting the extension via socket. You do **NOT** want to merge. I suggest to cherry pick the fix and write a different launch config. Since I am unfamiliar with your build system I did not change it but just disabled the build of the LS. I suggest you add an option to just build the extension (of course without bundling it).

I start the language server by just starting its main class `LanguageDiagramServer` and add `-Dport=5007` to the VM variables. You can find the corresponding solution [here](https://github.com/lf-lang/lingua-franca/tree/sdo/debug-ls).

**Reply by soerendomroes** [reply]
@petervdonovan This might be interesting for you.

**Reply by petervdonovan** [reply]
Thank you for this! I will try to get it working on my machine when I have time.
